core/src/banking_stage/transaction_scheduler/multi_iterator_consume_scheduler.rs

```diff
@@ -0,0 +1,617 @@
+use {
+    super::{
+        in_flight_tracker::InFlightTracker,
+        scheduler_error::SchedulerError,
+        thread_aware_account_locks::{ThreadAwareAccountLocks, ThreadId, ThreadSet},
+        transaction_priority_id::TransactionPriorityId,
+        transaction_state::SanitizedTransactionTTL,
+        transaction_state_container::TransactionStateContainer,
+    },
+    crate::banking_stage::{
+        consumer::TARGET_NUM_TRANSACTIONS_PER_BATCH,
+        multi_iterator_scanner::{
+            MultiIteratorScanner, PayloadAndAlreadyHandled, ProcessingDecision,
+        },
+        read_write_account_set::ReadWriteAccountSet,
+        scheduler_messages::{ConsumeWork, TransactionBatchId, TransactionId},
+    },
+    crossbeam_channel::Sender,
+    itertools::Itertools,
+    solana_sdk::{clock::Slot, transaction::SanitizedTransaction},
+};
+
+const QUEUED_TRANSACTION_LIMIT: usize = 64 * 100;
+
+/// Interface to perform scheduling for consuming transactions.
+/// Using a multi-iterator approach.
+pub struct MultiIteratorConsumeScheduler {
+    in_flight_tracker: InFlightTracker,
+    account_locks: ThreadAwareAccountLocks,
+    consume_work_senders: Vec<Sender<ConsumeWork>>,
+}
+
+impl MultiIteratorConsumeScheduler {
+    pub fn new(consume_work_senders: Vec<Sender<ConsumeWork>>) -> Self {
+        let num_threads = consume_work_senders.len();
+        Self {
+            in_flight_tracker: InFlightTracker::new(num_threads),
+            account_locks: ThreadAwareAccountLocks::new(num_threads),
+            consume_work_senders,
+        }
+    }
+
+    pub(crate) fn schedule(
+        &mut self,
+        container: &mut TransactionStateContainer,
+    ) -> Result<usize, SchedulerError> {
+        let num_threads = self.consume_work_senders.len();
+        let mut schedulable_threads = ThreadSet::any(num_threads);
+        let outstanding = self.in_flight_tracker.num_in_flight_per_thread();
+        for (thread_id, outstanding_count) in outstanding.iter().enumerate() {
+            if *outstanding_count > QUEUED_TRANSACTION_LIMIT {
+                schedulable_threads.remove(thread_id);
```

Comment:
Maybe add a test to demonstrate how thread are removed then add back to scheduler due to change of outstanding trnasactions.

Reply:
Have a test to test the removal, but no test to check if it was added back in.